### PR TITLE
Add support for accordion hints in tutorials

### PR DESCRIPTION
**Pull request by adrianwyatt**
This feature adds support for accordion hints in tutorial markdowns. 
[pxt-arcade issue 4555: Looking for a quick, simple solution to provide extra info without having to divert the coder's attention.](https://github.com/microsoft/pxt-arcade/issues/4555)

untitled

### CODE REVIEW

File: webapp/src/marked.tsx [hunk 1]
```diff
@@ -385,7 +486,7 @@ export class MarkedContent extends data.Component<MarkedContentProps, MarkedCont
         // Set markdown options
         marked.setOptions({
             renderer: renderer,
-            sanitize: true,
+            sanitize: false,
```
Comment (abchatra) [inline webapp/src/marked.tsx:489]: Early look, we can't set sanitize to false. This is a security issue with the third party injecting html. Why is this required?

Comment (adrianwyatt) [reply webapp/src/marked.tsx:489]: Ah - accidentally left that in there. I will set it back to "true".

Comment (abchatra) [reply webapp/src/marked.tsx:489]: Awesome thank you!

File: webapp/src/marked.tsx [hunk 2]
```diff
@@ -350,6 +352,105 @@ export class MarkedContent extends data.Component<MarkedContentProps, MarkedCont
...
+    // -------------------------------
+    private renderAccordianHints(content: HTMLElement) {
+        const hintBeginRegex = /^\s*~hint\s*(.+)/i;
+        const hintEndRegex = /^\s*hint~.*/i;
+
+        // Processing each hint modifies the node order and count of 'content.childnodes'.
+        // Keep reprocessing content.childnodes until all hints have been processed.
+        let hintFound: boolean = false;
+
+        do {
```
Comment (abchatra) [inline webapp/src/marked.tsx:379]: This is bit hard to read for me. Can we do this loop in two steps?

1. First loop finds all the hints
2. Second loop acts on the hints it found in the first loop.

Are we supporting nested hints here?

**Review by abchatra — CHANGES_REQUESTED** [review]

File: tests/tutorial-test/cases/accordionHints.md [hunk 3]
```diff
@@ -0,0 +1,10 @@
+# Accordion Hints
+## Step 1
+Welcome!
+
+~hint This content is hidden until the user clicks here.
```
Comment (abchatra) [inline tests/tutorial-test/cases/accordionHints.md:5]: Some more tests with the multiple hints and nested hints would be great. If we are not supporting nested hints we should call out (I am ok not supporting nested hints as well.

**Comment by abchatra** [comment]
Overall looks good to me.  I like ~hint & hint~ pair. However @kiki-lee is the expert here to confirm it is the right thing from an content author perspective.

@riknoll can you please review this as well?

File: theme/tutorial-sidebar.less [hunk 4]
```diff
@@ -434,6 +434,18 @@
     top: calc(5.5rem + @mainMenuHeight);
 }
 
+/*********************************
+        Accordion Hints (details)
+*********************************/
+
+.tabTutorial details {
+    padding: .5rem .5rem .5rem 2rem;
```
Comment (jwunderl) [inline theme/tutorial-sidebar.less:442]: 2rem left padding is pretty significant in the details view -- something like 10% of the whole width on my machine, which is notable when you're putting an image or block in a hint: 
image

```suggestion
    padding: .5rem;
```

feels a bit better to me personally:

image

File: webapp/src/marked.tsx [hunk 5]
```diff
@@ -350,6 +352,105 @@ export class MarkedContent extends data.Component<MarkedContentProps, MarkedCont
...
+            hintFound = false;
+            // The hint summary element (i.e. <summary>text</summary>).
+            let hintSummary: HTMLElement = null;
+            // The element that signifies the beginning of the hint (i.e. <p>~hint (hint summary)</p>).
```
Comment (jwunderl) [inline webapp/src/marked.tsx:383]: would be nice to have this logic included for docs view (e.g. localhost:3232/tutorials/chase-the-pizza) -- that logic lives over in https://github.com/microsoft/pxt/blob/master/pxtrunner/renderer.ts#L1302 . Currently it shows up as plaintext: 
image

(but this is also probably fine to handle separately / as a follow up, that view often ends up a little neglected to be honest and it's not the primary scenario)

File: docs/writing-docs/tutorials/basics.md [hunk 6]
```diff
@@ -168,6 +168,22 @@ forever(function() {
 ```
 ````
 
+## Accordion/hidden hints
+If you want to provide extra information without having to divert the coder's attention, you can include content in an "accordion" style hint control.
```
Comment (jwunderl) [inline docs/writing-docs/tutorials/basics.md:172]: It would maybe be worth dropping a link to https://makecode.com/writing-docs/tutorials/control-options#explicit-hints here? Or ideally images / snippets that are inside ~hints wouldn't trigger the 'hide in hint bubble' logic ever, but that may be a bit more complicated to implement.

Comment (kiki-lee) [reply docs/writing-docs/tutorials/basics.md:172]: Great point, Joey ^

File: tests/tutorial-test/cases/accordionHints.md (hunk 3 above)
Comment (kiki-lee) [reply tests/tutorial-test/cases/accordionHints.md:5]: > Some more tests with the multiple hints and nested hints would be great. If we are not supporting nested hints we should call out (I am ok not supporting nested hints as well.

I'm perfectly fine not supporting nested hints right now…that said, if we can get them, I'm sure they'll get used!  They'd just have to have slightly differentiated CSS so people could tell what's going on.

Comment (adrianwyatt) [reply tests/tutorial-test/cases/accordionHints.md:5]: I will add more test cases and keep nested accordion hints as not supported. I'll add a notice in the docs.

File: webapp/src/marked.tsx (hunk 2 above)
Comment (adrianwyatt) [reply webapp/src/marked.tsx]: We are not supporting nested hints. I have refactored the loops into two steps.

File: webapp/src/marked.tsx (hunk 5 above)
Comment (adrianwyatt) [reply webapp/src/marked.tsx]: I recommend handling this scenario in a separate PR.

File: docs/writing-docs/tutorials/basics.md [hunk 7]
```diff
@@ -168,6 +168,24 @@ forever(function() {
...
+
+### ~ hint
+Nested accordion hints are not currently supported.
+### ~
+
+```
+~hint This content is hidden until the user clicks here.
+  - :blank: Bullet 1
+  - :mouse pointer: Bullet 2
+hint~
```
Comment (adrianwyatt) [inline docs/writing-docs/tutorials/basics.md:186]: (not that a newline is no longer required)

File: webapp/src/marked.tsx (hunk 5 above)
Comment (jwunderl) [reply webapp/src/marked.tsx]: Yup, agreed

**Review by riknoll — APPROVED** [review]

**Comment by abchatra** [comment]
Congratulations!  ❤️
